content/blog/2021-08-24-pre-defined-models-exp.md

```diff
@@ -0,0 +1,374 @@
+
+In this post, we'll go through an example of fine-tuning
+[AlexNet](https://towardsdatascience.com/alexnet-the-architecture-that-challenged-cnns-e406d5297951)
+and
+[SqueezeNet](https://towardsdatascience.com/review-squeezenet-image-classification-e7414825581a)
+to classify bees and ants. We'll use DVC to handle experiments for us and we'll
+compare the results of both models at the end.
+
+## Initialize the pre-trained model
+
+We'll be fine-tuning the AlexNet model and the SqueezeNet model to classify
+images of bees and ants. You can find the project we're working with in
+[this repo](https://github.com/iterative/pretrained-model-demo), which is based
+on the tutorial over at
+[this post](https://pytorch.org/tutorials/beginner/finetuning_torchvision_models_tutorial.html).
+
+In the `pretrained_model_tuner.py` file, you'll find the code that defines both
+the AlexNet and SqueezeNet models. We start by initializing these models so we
+can get the number of model features and the input size we need for fine-tuning.
+
+Since the project has everything we need to initialize the models, we can start
+training and comparing the differences between them with the ants/bees dataset.
+Running experiments to get the best tuning for each model can make it difficult
+to see which changes led to a better result. That's why we will be using DVC to
+track changes in the code and the data.
+
+## Adding the train stage
+
+Stages in DVC let us define individual data processes and can be used to build
+detailed machine learning pipelines. You have the ability to define the
+different steps of model creation like preprocessing, featurization, and
+training.
+
+We currently have a `train` stage in the `dvc.yaml` file. If you take a look at
+it, you'll see something like:
+
+```yaml
+stages:
+  train:
+    cmd: python pretrained_model_tuner.py
+    deps:
+      - data/hymenoptera_data
+      - pretrained_model_tuner.py
+    params:
+      - lr
+      - momentum
+      - model_name
+      - num_classes
+      - batch_size
+      - num_epochs
+    outs:
+      - model.pt:
+          checkpoint: true
+    live:
+      results:
+        summary: true
+        html: true
+```
+
+The reason we need this `dvc.yaml` file is so DVC knows what to pay attention to
+in our workflow. It will start managing data, understand which metrics to pay
+attention to, and what the expected output for each step is.
+
+You'll typically add stages to `dvc.yaml` using the `dvc stage add` command and
+this is one of the ways you can add new stages or update existing ones.
+
+With the `train` stage defined, let's look at where the metrics actually come
+from in the code. If you open `pretrained_model_tuner`, you'll see a line where
+we dump the accuracy and loss for the training epochs into the `results.json`
+file. We're also saving the model on the epoch run and recording metrics for
+each epoch using `dvclive` logging.
+
+```python
+if phase == 'train':
```

Comment:
One more thought after taking a quick look at https://github.com/iterative/pretrained-model-demo/blob/132b6928ab11bbdb500e204c18e6491a03944b6b/pretrained_model_tuner.py#L84. The validation metrics are still being manually logged the old way and not captured by dvc. You could consider logging them with dvclive also like:

```python
            if phase == 'train':
                torch.save(model.state_dict(), "model.pt")

                dvclive.log('train_acc', epoch_acc.item())
                dvclive.log('train_loss', epoch_loss)
                dvclive.log('train_time', epoch_time_elapsed)

            if phase == 'val':
                dvclive.log('val_acc', epoch_acc.item())
                dvclive.log('val_loss', epoch_loss)
               
                dvclive.next_step()
```

Unfortunately, that does make your table even wider since you are adding more metrics. Another option might be to only log the validation metrics, since those are usually more helpful than training metrics, or to introduce the `--include-metrics`/`--include-params` flags to narrow the table.

Reply:
I added those metrics to the table and I took out a few of the params we aren't adjusting.

Reply:
Sounds good! Not sure if you are still making updates, but make sure to put `dvclive.next_step()` inside `if phase == 'val'` or else you will be making two steps for every epoch (once for train, once for val).

Reply:
Oh! That makes sense. I wasn't quite sure about that. I'll make those updates here soon!

Reply:
@dberenbaum Just pushed the update for this!